File: src/papyrus_api/routers/review.py
"""复习功能API路由。"""

from __future__ import annotations

from typing import Literal

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from papyrus.core import cards as card_core
from papyrus.core.cards import CardDict, NextDueResult
from papyrus_api.deps import get_data_file
from papyrus.paths import DATABASE_FILE
from papyrus.data.progress import record_card_reviewed

router = APIRouter(prefix="/review", tags=["review"])


class RateCardIn(BaseModel):
    grade: Literal[1, 2, 3] = Field(..., description="1=忘记, 2=模糊, 3=秒杀")


class NextDueEmptyResponse(BaseModel):
    success: bool
    card: None = None
    due_count: int
    total_count: int


class NextDueResponse(BaseModel):
    success: bool
    card: CardDict
    due_count: int
    total_count: int


class RateCardResponse(BaseModel):
    success: bool
    card: CardDict
    interval_days: float
    ef: float
    next: NextDueResult | None


@router.get("/next", response_model=NextDueResponse | NextDueEmptyResponse)
def get_next_due(tag: str | None = None) -> NextDueResponse | NextDueEmptyResponse:
    """获取下一张待复习的卡片。
    
    如果没有到期的卡片，返回空卡片和统计信息。
    """
    data_file = get_data_file()
    res = card_core.get_next_due(data_file, tag=tag)
    if res is None:
        cards = card_core.list_cards(data_file)
        return NextDueEmptyResponse(
            success=True,
            card=None,
            due_count=0,
            total_count=len(cards),
        )

    return NextDueResponse(success=True, **res)


@router.post("/{card_id}/rate", response_model=RateCardResponse)
def rate_card(card_id: str, payload: RateCardIn, tag: str | None = None) -> RateCardResponse:
    """对复习卡片进行评分。
    
    Args:
        card_id: 卡片ID
        payload.grade: 1=忘记, 2=模糊, 3=秒杀
    """
    data_file = get_data_file()
    res = card_core.rate_card(data_file, card_id, int(payload.grade))
    if res is None:
        raise HTTPException(status_code=404, detail="card not found")

    # 记录学习进度
    record_card_reviewed(DATABASE_FILE)

    # also return the next due snapshot for convenience
    nxt = card_core.get_next_due(data_file, tag=tag)
    return RateCardResponse(success=True, **res, next=nxt)

File: backend/src/api/routes/review.ts
import type { FastifyInstance } from 'fastify';
import { getNextDueCard, rateCard, getCardStats } from '../../core/cards.js';
import { recordCardReviewed } from '../../core/progress.js';
import { pushExtensionEvent } from '#/core/extension-events.js';

export default async function reviewRoutes(fastify: FastifyInstance): Promise<void> {
  fastify.get('/next', async (request, reply) => {
    try {
      const card = getNextDueCard();
      const stats = getCardStats();
      if (!card) {
        reply.send({
          success: true,
          card: null,
          due_count: stats.due,
          total_count: stats.total,
          message: 'No cards due for review',
        });
        return;
      }
      reply.send({
        success: true,
        card,
        due_count: stats.due,
        total_count: stats.total,
      });
    } catch (err) {
      const message = err instanceof Error ? err.message : '服务器内部错误';
      request.log.error({ err }, message);
      reply.status(500).send({ success: false, error: message });
    }
  });

  fastify.post('/:cardId/rate', async (request, reply) => {
    try {
      const { cardId } = request.params as { cardId: string };
      const body = request.body as { grade?: number };
      const grade = body.grade;

      if (grade !== 1 && grade !== 2 && grade !== 3) {
        reply.status(400).send({ success: false, error: 'Grade must be 1, 2, or 3' });
        return;
      }

      const result = await rateCard(cardId, grade);
      if (!result) {
        reply.status(404).send({ success: false, error: 'Card not found' });
        return;
      }
      recordCardReviewed();
      pushExtensionEvent('card.review.completed', {
        card_id: cardId,
        grade,
        interval_days: result.intervalDays,
        ef: result.ef,
      });

      const next = getNextDueCard();
      const stats = getCardStats();
      reply.send({
        success: true,
        card: result.card,
        interval_days: result.intervalDays,
        ef: result.ef,
        next: next ? {
          success: true,
          card: next,
          due_count: stats.due,
          total_count: stats.total,
        } : null,
      });
    } catch (err) {
      const message = err instanceof Error ? err.message : '服务器内部错误';
      request.log.error({ err }, message);
      reply.status(500).send({ success: false, error: message });
    }
  });
}
